// @it-healer/human-design-calculator/src/bodygraph.js
const { DateTime } = require('luxon');
const {
  initEphemeris, dateToJulianDay, julianDayToDate,
  getAllPositions, PLANET_NAMES,
} = require('./ephemeris');
const { getActivation, getOppositeActivation } = require('./activation');
const { findDesignDate } = require('./design-date');
const {
  getAllActivatedGates, getChannels, getDefinedCenters,
  getType, getTypeProperties, getAuthority, getDefinition,
  getProfile, getVariable, getCircuitries,
} = require('./type');
const {
  getIncarnationCross, getDetermination, getEnvironment,
  getView, getMotivation, getCognition, getSense,
} = require('./properties');
const { CENTER_NAMES } = require('./data/centers');

function julianDayToISO(jd) {
  const d = julianDayToDate(jd);
  const totalHours = d.hour;
  const hours = Math.floor(totalHours);
  const totalMinutes = (totalHours - hours) * 60;
  const minutes = Math.floor(totalMinutes);
  const seconds = Math.floor((totalMinutes - minutes) * 60);
  return DateTime.utc(d.year, d.month, d.day, hours, minutes, seconds).toISO();
}

function calculateBodygraph(dateStr, timeStr, tz, lat, lon, ephemerisPath) {
  initEphemeris(ephemerisPath);

  // 1. Convert local time → UTC
  const localDt = DateTime.fromISO(`${dateStr}T${timeStr}`, { zone: tz });
  if (!localDt.isValid) {
    throw new Error(`Invalid date/time/timezone: ${dateStr} ${timeStr} ${tz}`);
  }
  const utcDt = localDt.toUTC();

  // 2. Birth Julian Day
  const hourDecimal = utcDt.hour + utcDt.minute / 60 + utcDt.second / 3600;
  const birthJD = dateToJulianDay(utcDt.year, utcDt.month, utcDt.day, hourDecimal);

  // 3. Personality planetary positions and activations
  const personalityPositions = getAllPositions(birthJD);
  const personalityActs = {};
  for (const planet of PLANET_NAMES) {
    personalityActs[planet] = getActivation(personalityPositions[planet].longitude);
  }
  personalityActs['Earth'] = getOppositeActivation(personalityPositions.Sun.longitude);
  personalityActs['SouthNode'] = getOppositeActivation(personalityPositions.NorthNode.longitude);

  // 4. Find Design Date (88° retrograde sun)
  const designJD = findDesignDate(birthJD);

  // 5. Design planetary positions and activations
  const designPositions = getAllPositions(designJD);
  const designActs = {};
  for (const planet of PLANET_NAMES) {
    designActs[planet] = getActivation(designPositions[planet].longitude);
  }
  designActs['Earth'] = getOppositeActivation(designPositions.Sun.longitude);
  designActs['SouthNode'] = getOppositeActivation(designPositions.NorthNode.longitude);

  // 6. Derive gates, channels, centers
  const acts = { personality: personalityActs, design: designActs };
  const activatedGates = getAllActivatedGates(acts);
  const channels = getChannels(activatedGates);
  const definedCenters = getDefinedCenters(channels);

  // 7. Derive type, authority, definition
  const type = getType(channels, definedCenters, activatedGates);
  const { strategy, signature, notSelfTheme } = getTypeProperties(type);
  const authority = getAuthority(definedCenters, activatedGates);
  const definition = getDefinition(channels, definedCenters, activatedGates);
  const profile = getProfile(personalityActs.Sun.line, designActs.Sun.line);
  const variable = getVariable(
    personalityActs.Sun.tone, personalityActs.NorthNode.tone,
    designActs.Sun.tone, designActs.NorthNode.tone,
  );
  const circuitries = getCircuitries(channels);

  // 8. Derive properties
  const incarnationCross = getIncarnationCross(personalityActs.Sun, designActs.Sun);
  const determination = getDetermination(designActs.Sun);
  const environment = getEnvironment(designActs.NorthNode);
  const view = getView(personalityActs.NorthNode);
  const motivation = getMotivation(personalityActs.Sun);
  const cognition = getCognition(designActs.Sun);
  const sense = getSense(personalityActs.Sun);

  // 9. Center definition flags
  const centerDef = {};
  const defined = [];
  const undefined_ = [];
  for (const c of CENTER_NAMES) {
    const isDef = definedCenters.has(c);
    centerDef[c] = isDef;
    if (isDef) defined.push(c);
    else undefined_.push(c);
  }

  // 10. Assemble result
  return {
    input: {
      date: dateStr,
      time: timeStr,
      timezone: tz,
      latitude: lat,
      longitude: lon,
    },
    birthDateUTC: utcDt.toISO(),
    designDateUTC: julianDayToISO(designJD),
    activations: acts,
    type,
    strategy,
    signature,
    notSelfTheme,
    authority,
    definition,
    profile,
    incarnationCross,
    variable,
    channels,
    activatedGates: Array.from(activatedGates).sort((a, b) => a - b),
    definedCenters: defined,
    undefinedCenters: undefined_,
    centerDefinition: centerDef,
    circuitries,
    determination,
    environment,
    view,
    motivation,
    cognition,
    sense,
  };
}

module.exports = { calculateBodygraph };
