// @it-healer/human-design-calculator/src/ephemeris.js
const swisseph = require('swisseph');

const PLANETS = {
  Sun:       swisseph.SE_SUN,
  Moon:      swisseph.SE_MOON,
  Mercury:   swisseph.SE_MERCURY,
  Venus:     swisseph.SE_VENUS,
  Mars:      swisseph.SE_MARS,
  Jupiter:   swisseph.SE_JUPITER,
  Saturn:    swisseph.SE_SATURN,
  Uranus:    swisseph.SE_URANUS,
  Neptune:   swisseph.SE_NEPTUNE,
  Pluto:     swisseph.SE_PLUTO,
  NorthNode: swisseph.SE_TRUE_NODE,
};

const PLANET_NAMES = [
  'Sun', 'Moon', 'Mercury', 'Venus', 'Mars',
  'Jupiter', 'Saturn', 'Uranus', 'Neptune', 'Pluto', 'NorthNode'
];

const FLAG = swisseph.SEFLG_SPEED;

/** Initialize Swiss Ephemeris with path to .se1 files */
function initEphemeris(ephemerisPath) {
  swisseph.swe_set_ephe_path(ephemerisPath);
}

/** Convert date components to Julian Day */
function dateToJulianDay(year, month, day, hourDecimal) {
  let jd = 0;
  swisseph.swe_julday(year, month, day, hourDecimal, swisseph.SE_GREG_CAL,
    (julday) => { jd = julday; });
  return jd;
}

/** Convert Julian Day back to date components */
function julianDayToDate(jd) {
  let result = { year: 0, month: 0, day: 0, hour: 0 };
  swisseph.swe_revjul(jd, swisseph.SE_GREG_CAL,
    (res) => { result = res; });
  return result;
}

/** Get position of a single planet at a given Julian Day */
function getPlanetPosition(julianDay, planet) {
  let result = { longitude: 0, latitude: 0, distance: 0, longitudeSpeed: 0 };
  swisseph.swe_calc_ut(julianDay, planet, FLAG,
    (body) => {
      result = {
        longitude: body.longitude,
        latitude: body.latitude,
        distance: body.distance,
        longitudeSpeed: body.longitudeSpeed,
      };
    });
  return result;
}

/** Get all planetary positions for a given Julian Day */
function getAllPositions(julianDay) {
  const positions = {};
  for (const name of PLANET_NAMES) {
    positions[name] = getPlanetPosition(julianDay, PLANETS[name]);
  }
  return positions;
}

module.exports = {
  PLANETS, PLANET_NAMES,
  initEphemeris, dateToJulianDay, julianDayToDate,
  getPlanetPosition, getAllPositions,
};
