// @it-healer/human-design-calculator/src/properties.js
const {
  RIGHT_ANGLE_CROSSES,
  JUXTAPOSITION_CROSSES,
  LEFT_ANGLE_CROSSES,
} = require('./data/crosses');

/**
 * Incarnation Cross determination.
 */
function getIncarnationCross(pSun, dSun) {
  if (pSun.line < 5 && dSun.line >= 3) {
    return 'Right Angle Cross of ' + RIGHT_ANGLE_CROSSES[pSun.gate];
  } else if (pSun.line >= 5) {
    return 'Left Angle Cross of ' + LEFT_ANGLE_CROSSES[pSun.gate];
  } else {
    return 'Juxtaposition Cross of ' + JUXTAPOSITION_CROSSES[pSun.gate];
  }
}

const DETERMINATION_VALUES = ['Appetite', 'Taste', 'Thirst', 'Touch', 'Sound', 'Light'];
const ENVIRONMENT_VALUES = ['Caves', 'Markets', 'Kitchens', 'Mountains', 'Valleys', 'Shores'];
const VIEW_VALUES = ['Survival', 'Possibility', 'Power', 'Wanting', 'Probability', 'Personal'];
const MOTIVATION_VALUES = ['Fear', 'Hope', 'Desire', 'Need', 'Guilt', 'Innocence'];
const COGNITION_VALUES = ['Smell', 'Taste', 'Outer Vision', 'Inner Vision', 'Feeling', 'Touch'];
const SENSE_VALUES = ['Security', 'Uncertainty', 'Action', 'Meditation', 'Judgment', 'Acceptance'];

/** Determination: from Design Sun Color (1-6) */
function getDetermination(designSun) {
  return DETERMINATION_VALUES[designSun.color - 1];
}

/** Environment: from Design North Node Color (1-6) */
function getEnvironment(designNorthNode) {
  return ENVIRONMENT_VALUES[designNorthNode.color - 1];
}

/** View: from Personality North Node Color (1-6) */
function getView(personalityNorthNode) {
  return VIEW_VALUES[personalityNorthNode.color - 1];
}

/** Motivation: from Personality Sun Color (1-6) */
function getMotivation(personalitySun) {
  return MOTIVATION_VALUES[personalitySun.color - 1];
}

/** Cognition: from Design Sun Tone (1-6) */
function getCognition(designSun) {
  return COGNITION_VALUES[designSun.tone - 1];
}

/** Sense: from Personality Sun Tone (1-6) */
function getSense(personalitySun) {
  return SENSE_VALUES[personalitySun.tone - 1];
}

module.exports = {
  getIncarnationCross, getDetermination, getEnvironment,
  getView, getMotivation, getCognition, getSense,
};
