// @it-healer/human-design-calculator/src/type.js
const { HARMONIC_GATES } = require('./data/gates');
const { DEFINED_CENTERS_BY_CHANNEL, CIRCUITRY_BY_CHANNEL } = require('./data/channels');
const { BRIDGING_GATES } = require('./data/centers');

const TYPE_PROPERTIES = {
  'Manifestor':            { strategy: 'To Inform',                  signature: 'Peace',        notSelfTheme: 'Anger' },
  'Generator':             { strategy: 'To Respond',                 signature: 'Satisfaction',  notSelfTheme: 'Frustration' },
  'Manifesting Generator': { strategy: 'To Respond',                 signature: 'Satisfaction',  notSelfTheme: 'Frustration' },
  'Projector':             { strategy: 'To Wait for the Invitation', signature: 'Success',       notSelfTheme: 'Bitterness' },
  'Reflector':             { strategy: 'To Wait a Lunar Cycle',      signature: 'Surprise',      notSelfTheme: 'Disappointment' },
};

/** Get strategy, signature, and not-self theme for a given type */
function getTypeProperties(type) {
  return TYPE_PROPERTIES[type] || { strategy: '', signature: '', notSelfTheme: '' };
}

/** Collect all unique activated gate numbers from both sides */
function getAllActivatedGates(acts) {
  const gates = new Set();
  for (const a of Object.values(acts.personality)) gates.add(a.gate);
  for (const a of Object.values(acts.design)) gates.add(a.gate);
  return gates;
}

/** Derive active channels from activated gates using harmonic gate map */
function getChannels(activatedGates) {
  const channels = new Set();
  for (const gate of activatedGates) {
    const harmonics = HARMONIC_GATES[gate] || [];
    for (const h of harmonics) {
      if (activatedGates.has(h)) {
        const key = Math.min(gate, h) + '-' + Math.max(gate, h);
        channels.add(key);
      }
    }
  }
  return Array.from(channels).sort();
}

/** Derive defined centers from active channels */
function getDefinedCenters(channels) {
  const centers = new Set();
  for (const ch of channels) {
    const pair = DEFINED_CENTERS_BY_CHANNEL[ch];
    if (pair) {
      centers.add(pair[0]);
      centers.add(pair[1]);
    }
  }
  return centers;
}

/** Helper: check if a specific channel is active */
function hasChannel(activatedGates, gateA, gateB) {
  return activatedGates.has(gateA) && activatedGates.has(gateB);
}

/** Helper: check if any of the given gate pairs form a channel */
function hasAnyChannel(activatedGates, pairs) {
  return pairs.some(([a, b]) => hasChannel(activatedGates, a, b));
}

/**
 * Check if there is a motor-to-throat connection.
 * Motors: Solar Plexus, Sacral, Root, Ego.
 */
function motorToThroat(activatedGates, definedCenters) {
  if (!definedCenters.has('Throat')) return false;

  const sacralToThroat = hasChannel(activatedGates, 34, 20);
  const egoToThroat = hasChannel(activatedGates, 45, 21);
  const solarPlexusToThroat =
    hasChannel(activatedGates, 12, 22) || hasChannel(activatedGates, 35, 36);

  const gToThroat = hasAnyChannel(activatedGates, [[13, 33], [1, 8], [7, 31], [10, 20]]);
  const gToEgo = hasChannel(activatedGates, 51, 25);
  const gToSpleen = hasChannel(activatedGates, 10, 57);
  const gToSacral = hasAnyChannel(activatedGates, [[15, 5], [2, 14], [46, 29]]);

  const spleenToThroat =
    hasChannel(activatedGates, 16, 48) || hasChannel(activatedGates, 57, 20);
  const spleenToRoot = hasAnyChannel(activatedGates, [[54, 32], [28, 38], [18, 58]]);
  const spleenToEgo = hasChannel(activatedGates, 44, 26);

  return (
    sacralToThroat ||
    egoToThroat ||
    solarPlexusToThroat ||
    (gToThroat && gToEgo) ||
    (gToThroat && gToSpleen && spleenToRoot) ||
    (gToThroat && gToSacral) ||
    (spleenToThroat && spleenToRoot) ||
    (spleenToThroat && spleenToEgo)
  );
}

/** Determine Type */
function getType(channels, definedCenters, activatedGates) {
  if (channels.length === 0) return 'Reflector';
  const hasSacral = definedCenters.has('Sacral');
  const hasMotorToThroat = motorToThroat(activatedGates, definedCenters);
  if (hasSacral) return hasMotorToThroat ? 'Manifesting Generator' : 'Generator';
  return hasMotorToThroat ? 'Manifestor' : 'Projector';
}

/**
 * Determine Authority.
 */
function getAuthority(definedCenters, activatedGates) {
  if (definedCenters.size === 0) return 'Lunar';
  if (definedCenters.has('Solar Plexus')) return 'Solar Plexus';
  if (definedCenters.has('Sacral')) return 'Sacral';
  if (definedCenters.has('Spleen')) return 'Spleen';

  // Ego authority
  const egoToThroat = hasChannel(activatedGates, 45, 21);
  const gToEgo = hasChannel(activatedGates, 51, 25);
  if (egoToThroat || gToEgo) return 'Ego';

  // Self Projected
  const gToThroat = hasAnyChannel(activatedGates, [[13, 33], [1, 8], [7, 31], [10, 20]]);
  if (gToThroat) return 'Self Projected';

  // Outer Authority (Mental Projector)
  const headToAjna = hasAnyChannel(activatedGates, [[64, 47], [61, 24], [63, 4]]);
  const ajnaToThroat = hasAnyChannel(activatedGates, [[17, 62], [43, 23], [11, 56]]);
  if (headToAjna || ajnaToThroat) return 'Outer Authority';

  return 'Lunar';
}

/**
 * Determine Definition using union-find algorithm.
 */
function getDefinition(channels, definedCenters, activatedGates) {
  if (channels.length === 0) return 'None';

  const parent = {};
  for (const c of definedCenters) parent[c] = c;

  function find(x) {
    while (parent[x] !== x) {
      parent[x] = parent[parent[x]];
      x = parent[x];
    }
    return x;
  }

  function union(a, b) {
    const ra = find(a);
    const rb = find(b);
    if (ra !== rb) parent[ra] = rb;
  }

  for (const ch of channels) {
    const pair = DEFINED_CENTERS_BY_CHANNEL[ch];
    if (pair) union(pair[0], pair[1]);
  }

  const roots = new Set();
  for (const c of definedCenters) roots.add(find(c));
  const areaCount = roots.size;

  if (areaCount === 1) return 'Single';
  if (areaCount === 3) return 'Triple Split';
  if (areaCount >= 4) return 'Quad Split';

  // areaCount === 2: check for Wide Split
  const areas = {};
  for (const c of definedCenters) {
    const root = find(c);
    if (!areas[root]) areas[root] = [];
    areas[root].push(c);
  }
  const [areaA, areaB] = Object.values(areas);

  for (const centerA of areaA) {
    for (const centerB of areaB) {
      const bridges = BRIDGING_GATES[centerA] && BRIDGING_GATES[centerA][centerB];
      if (bridges) {
        for (const gate of bridges) {
          if (activatedGates.has(gate)) {
            return 'Split';
          }
        }
      }
    }
  }

  return 'Wide Split';
}

/** Profile from Personality Sun Line / Design Sun Line */
function getProfile(pSunLine, dSunLine) {
  return `${pSunLine}/${dSunLine}`;
}

/** Variable: P[L/R][L/R] D[L/R][L/R] based on Tone values */
function getVariable(pSunTone, pNodeTone, dSunTone, dNodeTone) {
  return 'P'
    + (pSunTone < 4 ? 'L' : 'R')
    + (pNodeTone < 4 ? 'L' : 'R')
    + ' D'
    + (dSunTone < 4 ? 'L' : 'R')
    + (dNodeTone < 4 ? 'L' : 'R');
}

/** Unique circuitries from active channels */
function getCircuitries(channels) {
  const set = new Set();
  for (const ch of channels) {
    const c = CIRCUITRY_BY_CHANNEL[ch];
    if (c) set.add(c);
  }
  return Array.from(set).sort();
}

module.exports = {
  getTypeProperties, getAllActivatedGates, getChannels, getDefinedCenters,
  motorToThroat, getType, getAuthority, getDefinition,
  getProfile, getVariable, getCircuitries,
};
